### PR TITLE
IndexManager to manage control of indexing into separate Solr URLs.

### PR DESCRIPTION
**Pull request by tpendragon**
Work towards #1518 

- [x] Rake task to create an IndexManager on first deploy which has the solr URL and the most recently created dump.
- [x] Rake task/class method to do a full reindex.

### CODE REVIEW

**Comment by tpendragon** [comment]
The full reindex logic may not be correct. The first incremental is a full index, as is the first "full index". Do we have some way of identifying an incremental which is everything?

File: spec/models/index_manager_spec.rb [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+      # There's one record in 1.xml, and one record in 2.xml
+      expect(response['response']['numFound']).to eq 2
+      index_manager.reload
+      expect(index_manager.last_dump_completed).to eq full_event.dump
+      expect(index_manager.dump_in_progress).to be_nil
+      expect(index_manager).not_to be_in_progress
+    end
+    it "doesn't index anything if it's caught up" do
+      allow(DumpFileIndexJob).to receive(:perform_async).and_call_original
+      # This incremental is before the full dump, don't run it
```
Comment (hackartisan) [inline spec/models/index_manager_spec.rb:79]: Not sure this comment is accurate?

File: spec/models/index_manager_spec.rb [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        index_manager = described_class.find(index_manager.id)
+        index_manager.index_next_dump!
+        run_callback(Sidekiq::BatchSet.new.to_a.last)
+      end
+      solr.commit
+
+      expect(DumpFileIndexJob).not_to have_received(:perform_async).with(pre_pre_incremental_event.dump.dump_files.first.id, anything)
+      expect(DumpFileIndexJob).to have_received(:perform_async).with(pre_incremental_event.dump.dump_files.first.id, anything)
+      expect(DumpFileIndexJob).to have_received(:perform_async).with(incremental_event.dump.dump_files.first.id, anything)
+      expect(DumpFileIndexJob).not_to have_received(:perform_async).with(final_incremental_event.dump.dump_files.first.id, anything)
```
Comment (hackartisan) [inline spec/models/index_manager_spec.rb:126]: This one doesn't run just because you didn't run the callbacks manually enough times?

**Review by hackartisan — CHANGES_REQUESTED** [review]
This is really awesome and exciting!

One minor fix inline.

My current thoughts are around
- what it looks like to swap an index - the index managers will have the alias urls so they'll need to be updated.
- what updates we will need to documentation

But none of that needs to be addressed yet in this PR

File: spec/models/index_manager_spec.rb (hunk 2 above)
Comment (tpendragon) [reply spec/models/index_manager_spec.rb:126]: Yes.

**Review by hackartisan — APPROVED** [review]

**Pull request merged**